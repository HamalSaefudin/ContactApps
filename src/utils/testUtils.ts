import {Reducer, createStore} from 'redux';

export const globalMockReducers = (store: Reducer) => {
  return createStore(store);
};
