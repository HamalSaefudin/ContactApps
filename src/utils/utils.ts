export const getImg = (url: string): string => {
  if (url.startsWith('http://') || url.startsWith('https://')) {
    return url.startsWith('http://') ? url.replace('http://', 'https://') : url;
  } else {
    return 'https://encrypted-tbn0.gstatic.com/images?q=tbn:ANd9GcQ4YreOWfDX3kK-QLAbAL4ufCPc84ol2MA8Xg&s';
  }
};
