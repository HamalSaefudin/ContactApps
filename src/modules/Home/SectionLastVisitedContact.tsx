import {
  FlatList,
  ListRenderItemInfo,
  StyleSheet,
  Text,
  View,
} from 'react-native';
import React, {useCallback} from 'react';
import {useSelector} from 'react-redux';
import {moderateScale} from 'react-native-size-matters';
import {Image} from 'react-native';
import {IContact} from '@/types/contact';
import Layout from '@/theme/Layout';
import Colors from '@/theme/Colors';
import {getImg} from '@/utils/utils';
import Fonts from '@/theme/Fonts';

export default function SectionLastVisitedContact() {
  const {lastVisitContact} = useSelector((state: any) => state.contacts);

  const renderLastVisitedContact = useCallback(
    (prop: ListRenderItemInfo<IContact>) => {
      return (
        <View style={Layout.alignItemsCenter}>
          <View style={styles.lastVisitedContactContainer}>
            <Image
              source={{uri: getImg(prop.item.photo)}}
              style={[Layout.fullSize, {borderRadius: moderateScale(50 / 2)}]}
            />
          </View>
          <Text>{`${prop.item.firstName} ${prop.item.lastName}`}</Text>
        </View>
      );
    },
    [],
  );

  if (lastVisitContact.length === 0) {
    return null;
  }

  return (
    <>
      <Text style={Fonts.headingMedium.h3}>Last Visited Contact</Text>
      <FlatList
        testID="list-last-visited-contact"
        data={lastVisitContact}
        horizontal
        style={{marginVertical: moderateScale(16)}}
        contentContainerStyle={{gap: moderateScale(20)}}
        showsHorizontalScrollIndicator={false}
        keyExtractor={(_, index) => index.toString()}
        renderItem={renderLastVisitedContact}
      />
      <View style={styles.border} />
    </>
  );
}

const styles = StyleSheet.create({
  lastVisitedContactContainer: {
    height: moderateScale(50),
    width: moderateScale(50),
    borderRadius: moderateScale(50) / 2,
  },
  border: {
    ...Layout.fullWidth,
    height: 1,
    backgroundColor: Colors.common.border,
  },
});
