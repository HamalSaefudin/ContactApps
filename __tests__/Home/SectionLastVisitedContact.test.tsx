import SectionLastVisitedContact from '@/modules/Home/SectionLastVisitedContact';
import {globalMockReducers} from '@/utils/testUtils';
import {render} from '@testing-library/react-native';
import React from 'react';
import {Provider} from 'react-redux';
import {combineReducers} from 'redux';

describe('SectionLastVisitedContact', () => {
  let store: any;

  beforeEach(() => {
    store = globalMockReducers(
      combineReducers({
        contacts: () => ({
          lastVisitContact: [
            {
              id: '1',
              firstName: 'John',
              lastName: 'Doe',
              age: 30,
              photo: 'https://example.com/photo.jpg',
              callHistory: '2024-06-09 12:00:00',
            },
            {
              id: '2',
              firstName: 'Jane',
              lastName: 'Smith',
              age: 25,
              photo: 'https://example.com/photo2.jpg',
              callHistory: '2024-06-09 12:00:00',
            },
          ],
        }),
      }),
    );
  });

  it('to match snapshot ', () => {
    const {toJSON} = render(
      <Provider store={store}>
        <SectionLastVisitedContact />
      </Provider>,
    );
    expect(toJSON()).toMatchSnapshot();
  });
  it('renders correctly ', () => {
    const {getByText, getByTestId} = render(
      <Provider store={store}>
        <SectionLastVisitedContact />
      </Provider>,
    );

    expect(getByText('Last Visited Contact')).toBeTruthy();
    expect(getByText('John Doe')).toBeTruthy();
    expect(getByText('Jane Smith')).toBeTruthy();

    const flatList = getByTestId('list-last-visited-contact');
    expect(flatList.props.data.length).toBe(2);
  });

  it('renders no contacts', () => {
    store = globalMockReducers(
      combineReducers({
        contacts: () => ({
          lastVisitContact: [],
        }),
      }),
    );

    const {queryByText} = render(
      <Provider store={store}>
        <SectionLastVisitedContact />
      </Provider>,
    );

    expect(queryByText('Last Visited Contact')).not.toBeTruthy();
  });
});
