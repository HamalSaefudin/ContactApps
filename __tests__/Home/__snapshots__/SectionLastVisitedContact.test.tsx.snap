// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SectionLastVisitedContact to match snapshot  1`] = `
[
  <Text>
    Last Visited Contact
  </Text>,
  <RCTScrollView
    contentContainerStyle={
      {
        "gap": 31.428571428571427,
      }
    }
    data={
      [
        {
          "age": 30,
          "callHistory": "2024-06-09 12:00:00",
          "firstName": "John",
          "id": "1",
          "lastName": "Doe",
          "photo": "https://example.com/photo.jpg",
        },
        {
          "age": 25,
          "callHistory": "2024-06-09 12:00:00",
          "firstName": "Jane",
          "id": "2",
          "lastName": "Smith",
          "photo": "https://example.com/photo2.jpg",
        },
      ]
    }
    getItem={[Function]}
    getItemCount={[Function]}
    horizontal={true}
    keyExtractor={[Function]}
    onContentSizeChange={[Function]}
    onLayout={[Function]}
    onMomentumScrollBegin={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={0.0001}
    showsHorizontalScrollIndicator={false}
    stickyHeaderIndices={[]}
    style={
      {
        "marginVertical": 25.142857142857142,
      }
    }
    testID="list-last-visited-contact"
    viewabilityConfigCallbackPairs={[]}
  >
    <View>
      <View
        onFocusCapture={[Function]}
        onLayout={[Function]}
        style={
          [
            {
              "flexDirection": "row",
            },
            null,
          ]
        }
      >
        <View
          style={
            {
              "alignItems": "center",
            }
          }
        >
          <View
            style={
              {
                "borderRadius": 39.285714285714285,
                "height": 78.57142857142857,
                "width": 78.57142857142857,
              }
            }
          >
            <Image
              source={
                {
                  "uri": "https://example.com/photo.jpg",
                }
              }
              style={
                [
                  {
                    "height": "100%",
                    "width": "100%",
                  },
                  {
                    "borderRadius": 39.285714285714285,
                  },
                ]
              }
            />
          </View>
          <Text>
            John Doe
          </Text>
        </View>
      </View>
      <View
        onFocusCapture={[Function]}
        onLayout={[Function]}
        style={
          [
            {
              "flexDirection": "row",
            },
            null,
          ]
        }
      >
        <View
          style={
            {
              "alignItems": "center",
            }
          }
        >
          <View
            style={
              {
                "borderRadius": 39.285714285714285,
                "height": 78.57142857142857,
                "width": 78.57142857142857,
              }
            }
          >
            <Image
              source={
                {
                  "uri": "https://example.com/photo2.jpg",
                }
              }
              style={
                [
                  {
                    "height": "100%",
                    "width": "100%",
                  },
                  {
                    "borderRadius": 39.285714285714285,
                  },
                ]
              }
            />
          </View>
          <Text>
            Jane Smith
          </Text>
        </View>
      </View>
    </View>
  </RCTScrollView>,
  <View
    style={
      {
        "backgroundColor": "#D9D9D9",
        "height": 1,
        "width": "100%",
      }
    }
  />,
]
`;
